### DISCUSSION TITLE
NC to DFS2

### DISCUSSION THREAD

**Discussion by pradhancda** [DHI/mikeio#513] [Q&A]
Sir,

Would you please suggest a possible solution to this error. 

Traceback (most recent call last):
  File "E:\testing_automation\batch_scripts\ncmrwf_data_dfs2.py", line 63, in <module>
    dfs.write(filename=dfsfilename,
TypeError: Dfs2.write() got an unexpected keyword argument 'start_time'

thanking you,
Subhasis

**Comment by jsmariegaard** [comment]
Hi @pradhancda, could you let us know which version of MIKE IO you're using? And show a bit more of the script? 

The error message "Dfs2.write() got an unexpected keyword argument 'start_time'" looks correct to me as this method has no argument "start_time".

**Reply by pradhancda** [reply]
Sir,
I am using mikeio 1.2.5
Please find my code and suggest the possible solutions.

....................my code ..................................
from datetime import datetime, timedelta
import xarray
import numpy as np
import pandas as pd
from mikeio import Dfs2

file = f"ncmrwf_2022.nc"
ds = xarray.open_dataset (file)

lat = ds.latitude.values
lon = ds.longitude.values
nx = len(lon)
ny = len(lat)
x0 = lon[0]
y0 = lat[0]
dx = np.round((lon[-1] - lon[0]) / nx,2)
dy = np.round((lat[-1] - lat[0]) / ny,2)
x0, y0, nx, ny, dx, dy

t = ds.time.values
start_time = pd.to_datetime(t).to_pydatetime()[0]

ds = ds.isel(time=slice(0,12))
rh = ds.RH_2m.values
u = ds.UGRD_10m.values
v = ds.VGRD_10m.values
temp = ds.TMP_2m.values
nlwrf = ds.NLWRF_sfc.values
nswrf = ds.NSWRF_sfc.values
mslp = ds.PRMSL_msl.values / 100

from mikeio.eum import ItemInfo, EUMUnit, EUMType

d = [rh, mslp, temp, u, v, nlwrf, nswrf]

dfsfilename = f"ncmwrwf_2022.dfs2"

coordinate = ['LONG/LAT', x0, y0, 0]

items = [ItemInfo("Relative Humidity", EUMType.Relative_humidity, EUMUnit.percent),
         ItemInfo("Mean Sea Level Pressure", EUMType.Air_Pressure, EUMUnit.hectopascal),
         ItemInfo("Air temperature at 2M", EUMType.Temperature, EUMUnit.degree_Kelvin),
         ItemInfo("Wind U", EUMType.Wind_Velocity, EUMUnit.meter_per_sec),
         ItemInfo("Wind V", EUMType.Wind_Velocity, EUMUnit.meter_per_sec),
         ItemInfo("Net Longwave Radiation Flux", EUMType.Heat_Flux, EUMUnit.watt_per_meter_pow_2),
         ItemInfo("Net Short Radiation Flux", EUMType.Heat_Flux, EUMUnit.watt_per_meter_pow_2)]
dfs = Dfs2()
dfs.write(filename=dfsfilename,
           data=d,
           start_time = start_time,
           dt=21600,
           items=items,
           coordinate=coordinate, dx=dx, dy=dy
)
.........................................................................................................

regards,
Subhasis

**Reply by ecomodeller** [reply]
Take a look at this notebook https://dhi.github.io/getting-started-with-mikeio/netcdf.html#convert-to-dfs2

**Reply by pradhancda** [reply]
Sir,
Thank you.

I followed the above link.
However, the converted dfs2 file shows opposite (N-S). 
when i tried to reverse the coordinates (g = mikeio.Grid2D(x=ds.longitude, y=ds.latitude[::-1], projection="Lon/Lat")), 
gives following error

**error message**
ValueError                                Traceback (most recent call last)
Cell In [220], line 6
      2 time
      4 #g = mikeio.Grid2D(x=ds.longitude.values, y=ds.latitude.values, projection="LONG/LAT")
      5 #g = mikeio.Grid2D(x=ds.longitude, y=ds.latitude, projection="LONG/LAT")
----> 6 g = mikeio.Grid2D(x=ds.longitude, y=ds.latitude[::-1], projection="Lon/Lat")
      7 g

File c:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\mikeio\spatial\grid_geometry.py:328, in Grid2D.__init__(self, x, x0, dx, nx, y, y0, dy, ny, bbox, projection, origin, orientation, axis_names, is_spectral)
    326     self._x0, self._dx, self._nx = _parse_grid_axis("x", x, x0, dx, nx)
    327     dy = self._dx if dy is None else dy
--> 328     self._y0, self._dy, self._ny = _parse_grid_axis("y", y, y0, dy, ny)
    330 self.is_spectral = is_spectral
    332 self.plot = _Grid2DPlotter(self)

File c:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\mikeio\spatial\grid_geometry.py:26, in _parse_grid_axis(name, x, x0, dx, nx)
     24 _check_equidistant(x)
     25 if len(x) > 1 and x[0] > x[-1]:
---> 26     raise ValueError(f"{name} values must be increasing")
     27 x0 = x[0]
     28 dx = x[1] - x[0] if len(x) > 1 else 1.0

ValueError: y values must be increasing

![mikeio_nc2dfs2](https://user-images.githubusercontent.com/48046958/218406993-7744bf02-54c7-4a57-9e1d-57ba4b422081.png)

netCDF data link
https://we.tl/t-eHZjxov4pd 

Code file
[code.txt](https://github.com/DHI/mikeio/files/10720158/code.txt)


Sir, Please suggest to rectify this error.

thank yoy

regards,

Subhasis Pradhan

**Reply by jsmariegaard** [reply]
Hi - you need to flip the data upside down not the axis.

**Reply by pradhancda** [reply]
Sir, 
 can i get one example ?

please...


regards,
Subhasis

**Reply by pradhancda** [reply]
Sir, 

Thank you so much.

I got success.

regards, 
Subhasis

**Reply by jsmariegaard** [reply]
Great :-)

**Reply by pradhancda** [reply]
Sir,

Again, I faced a problem during converting another wind file from nc to dfs2.

Output exceeds the [size limit](command:workbench.action.openSettings?[). Open the full output data [in a text editor](command:workbench.action.openLargeOutput?1b4fee4a-f06a-4117-a6e9-fbd3e3f52521)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In [43], line 5
      1 gu = mikeio.Grid2D(x=dsu.longitude.values, y=dsu.latitude.values, projection="LONG/LAT")
      2 gv = mikeio.Grid2D(x=dsv.longitude.values, y=dsv.latitude.values, projection="LONG/LAT")
      4 das = [
----> 5     mikeio.DataArray(data=dsu.eastward_wind.values,
      6                      time=timeu, geometry=gu,
      7                      item=mikeio.ItemInfo("U_Wind", mikeio.EUMType.Wind_Velocity, mikeio.EUMUnit.meter_per_sec)),
      8     mikeio.DataArray(data=dsv.northward_wind.values,
      9                      time=timev, geometry=gv,
     10                      item=mikeio.ItemInfo("V_Wind", mikeio.EUMType.Wind_Velocity, mikeio.EUMUnit.meter_per_sec))                 
     11 ]
     12 my_ds = mikeio.Dataset(das)
     14 my_ds.to_dfs("ascat_wind_Raka.dfs2")

File c:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\mikeio\dataarray.py:718, in DataArray.__init__(self, data, time, item, geometry, zn, dims)
    715 self._check_time_data_length(self.time)
    717 self.item = self._parse_item(item)
--> 718 self.geometry = self._parse_geometry(geometry, self.dims, self.shape)
    719 self._zn = self._parse_zn(zn, self.geometry, self.n_timesteps)
    720 self._set_spectral_attributes(geometry)

File c:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\mikeio\dataarray.py:853, in DataArray._parse_geometry(geometry, dims, shape)
    849     assert (
...
--> 853     assert shape[axis] == geometry.ny, "data shape does not match ny"
    854     assert shape[axis + 1] == geometry.nx, "data shape does not match nx"
    855 # elif isinstance(geometry, Grid3D): # TODO

AssertionError: data shape does not match ny

[uwind_raka.txt](https://github.com/DHI/mikeio/files/10730776/uwind_raka.txt)

Sir... Please suggest the possible solution.

regards,
Subhasis
